src/Npgsql/NpgsqlConnector.cs

```diff
@@ -494,10 +500,38 @@ internal async Task LoadDatabaseInfo(bool forceReload, NpgsqlTimeout timeout, bo
             // being set up (even if its empty)
             TypeMapper = new ConnectorTypeMapper(this);
 
-            if (forceReload || !NpgsqlDatabaseInfo.Cache.TryGetValue(ConnectionString, out var database))
-                NpgsqlDatabaseInfo.Cache[ConnectionString] = database = await NpgsqlDatabaseInfo.Load(Connection, timeout, async);
+            NpgsqlDatabaseInfo? database;
+
+            if (forceReload)
+            {
+                NpgsqlDatabaseInfo.Cache[ConnectionString] = database = await NpgsqlDatabaseInfo.Load(Connection,
+                    timeout, async);
+            }
+            else if (!NpgsqlDatabaseInfo.Cache.TryGetValue(ConnectionString, out database))
+            {
+                var hasSemaphore = async
+                    ? await DatabaseInfoSemaphore.WaitAsync(timeout.TimeLeft, cancellationToken)
+                    : DatabaseInfoSemaphore.Wait(timeout.TimeLeft, cancellationToken);
+
+                // We've timed out - calling Check, to throw the correct exception
+                if (!hasSemaphore)
+                    timeout.Check();
```

Comment:
Should be inside of the `try-finally` block,. Otherwise, all info loading will be locked forever after the first timeout.

Reply:
Nope - we get here only if there was a timeout while taking the semaphore.

Reply:
Missed the condition, right.